### PR TITLE
ddl:`show full columns` for date, datetime, timestamp compatibility.

### LINKED ISSUE #9807: compatibility: “show full columns” prints wrong Collation for date, datatime, timestamp columns
```
create table t1 (a date,b datetime,c timestamp);
show full columns from t1;
```
TiDB shows a different collation value ("binary") than mysql(null), this is not a big deal in most cases. It's reported that a CDC software called "Attunity" check the "collation" output of "show full columns" of the tables that managed by itself.
the tool can't pass the db usability check ,pls take a look.

MySQL 5.7.20

> +-------+-----------+-----------+------+-----+---------+-------+---------------------------------+---------+
> | Field | Type      | Collation | Null | Key | Default | Extra | Privileges                      | Comment |
> +-------+-----------+-----------+------+-----+---------+-------+---------------------------------+---------+
> | a     | date      | NULL      | YES  |     | NULL    |       | select,insert,update,references |         |
> | b     | datetime  | NULL      | YES  |     | NULL    |       | select,insert,update,references |         |
> | c     | timestamp | NULL      | YES  |     | NULL    |       | select,insert,update,references |         |
> +-------+-----------+-----------+------+-----+---------+-------+---------------------------------+---------+
> 3 rows in set (0.00 sec)

TiDB 2.1.6:

> +-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
> | Field | Type      | Collation | Null | Key  | Default | Extra | Privileges                      | Comment |
> +-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
> | a     | date      | binary    | YES  |      | NULL    |       | select,insert,update,references |         |
> | b     | datetime  | binary    | YES  |      | NULL    |       | select,insert,update,references |         |
> | c     | timestamp | binary    | YES  |      | NULL    |       | select,insert,update,references |         |
> +-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
> 3 rows in set (0.01 sec)

> Release Version: v2.1.6
> Git Commit Hash: 66e639e8e127f71662f37287eb984dd84d169cba
> Git Branch: HEAD
> UTC Build Time: 2019-03-15 09:09:07
> GoVersion: go version go1.11.2 linux/amd64
> Race Enabled: false
> TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
> Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by xiekeyi98**
### What problem does this PR solve? 

Fix #9807 
~~Relate PR: https://github.com/pingcap/parser/pull/249~~

### What is changed and how it works?

When we add a new column, we check its type and set it NULL.

### Check List 

Tests 

 - Unit test
 - Integration test 
 - Manual test

### CODE REVIEW

**Comment by xiekeyi98** [comment]
```SQL
mysql> create table t1 (a date,b datetime,c timestamp);
Query OK, 0 rows affected (0.01 sec)
mysql> show tables;
+----------------+
| Tables_in_test |
+----------------+
| t1             |
+----------------+
1 row in set (0.00 sec)
mysql> show full columns from t1;
+-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
| Field | Type      | Collation | Null | Key  | Default | Extra | Privileges                      | Comment |
+-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
| a     | date      | NULL      | YES  |      | NULL    |       | select,insert,update,references |         |
| b     | datetime  | NULL      | YES  |      | NULL    |       | select,insert,update,references |         |
| c     | timestamp | NULL      | YES  |      | NULL    |       | select,insert,update,references |         |
+-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
3 rows in set (0.00 sec)
mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v3.0.0-beta-242-g0b28f30cf-dirty
Git Commit Hash: 0b28f30cfdef7a396c7d5474d5536787e2d00bd0
Git Branch: issue9807
UTC Build Time: 2019-03-19 11:23:47
GoVersion: go version go1.12 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
1 row in set (0.00 sec)
```

**Comment by xiekeyi98** [comment]
CI will be fixed after https://github.com/pingcap/parser/pull/249 merged.

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -234,6 +234,9 @@ func setCharsetCollationFlenDecimal(tp *types.FieldType, tblCharset string, dbCh
 			if err != nil {
 				return errors.Trace(err)
 			}
+		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
+			tp.Collate = charset.NULL
```
Comment (kennytm) [inline ddl/ddl_api.go]:
```suggestion
			tp.Charset = charset.NULL
```

Comment (xiekeyi98) [reply ddl/ddl_api.go]: Extremely sorry for my foolish.

**Comment by shenli** [comment]
There is no unit test in this PR.

**Comment by xiekeyi98** [comment]
Sorry for my fault.
Changing the charset when creating needs to change many codes, I consider to only change when show columns.

File: executor/show_test.go [hunk 2]
```diff
@@ -191,7 +198,7 @@ func (s *testSuite2) TestShow2(c *C) {
 	tk.Se.Auth(&auth.UserIdentity{Username: "root", Hostname: "192.168.0.1", AuthUsername: "root", AuthHostname: "%"}, nil, []byte("012345678901234567890"))
 
 	r := tk.MustQuery("show table status from test like 't'")
-	r.Check(testkit.Rows(fmt.Sprintf("t InnoDB 10 Compact 0 0 0 0 0 0 0 %s <nil> <nil> utf8mb4_bin   注释", createTime)))
+	r.Check(testkit.Rows(fmt.Sprintf("t InnoDB 10 Compact 0 0 0 0 0 0 0 %s <nil> <nil> utf8mb4_bin   ", createTime)))
```
Comment (xiekeyi98) [inline executor/show_test.go]: I'm puzzled for this line.
But if I do not delete `注释`, it cannot pass the test.
However, I think what I changed should not influence this case.
```
FAIL: show_test.go:161: testSuite2.TestShow2

show_test.go:201:
    r.Check(testkit.Rows(fmt.Sprintf("t InnoDB 10 Compact 0 0 0 0 0 0 0 %s <nil> <nil> utf8mb4_bin   注释", createTime)))
/home/jenkins/workspace/tidb_ghpr_check/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:55:
    res.c.Assert(resBuff.String(), check.Equals, needBuff.String(), res.comment)
... obtained string = "[t InnoDB 10 Compact 0 0 0 0 0 0 0 2019-03-20 12:54:49 <nil> <nil> utf8mb4_bin   ]\n"
... expected string = "[t InnoDB 10 Compact 0 0 0 0 0 0 0 2019-03-20 12:54:49 <nil> <nil> utf8mb4_bin   注释]\n"
... sql:show table status from test like 't', args:[]
```

Comment (xiekeyi98) [reply executor/show_test.go]: I have got it.

File: executor/show.go [hunk 3]
```diff
@@ -390,6 +390,10 @@ func (e *ShowExec) fetchShowColumns() error {
 		// The FULL keyword causes the output to include the column collation and comments,
 		// as well as the privileges you have for each column.
 		if e.Full {
+			switch desc.Type {
+			case "date", "datetime", "timestamp":
```
Comment (crazycs520) [inline executor/show.go]: I'm afraid all time type have this  problem. And note the decimal.
In mysql 
```sql
> create table t1 (a date,b datetime,c timestamp,d time, e year);
> show full columns from t1;
+-------+-----------+-----------+------+-----+-------------------+-----------------------------+---------------------------------+---------+
| Field | Type      | Collation | Null | Key | Default           | Extra                       | Privileges                      | Comment |
+-------+-----------+-----------+------+-----+-------------------+-----------------------------+---------------------------------+---------+
| a     | date      | <null>    | YES  |     | <null>            |                             | select,insert,update,references |         |
| b     | datetime  | <null>    | YES  |     | <null>            |                             | select,insert,update,references |         |
| c     | timestamp | <null>    | NO   |     | CURRENT_TIMESTAMP | on update CURRENT_TIMESTAMP | select,insert,update,references |         |
| d     | time      | <null>    | YES  |     | <null>            |                             | select,insert,update,references |         |
| e     | year(4)   | <null>    | YES  |     | <null>            |                             | select,insert,update,references |         |
+-------+-----------+-----------+------+-----+-------------------+-----------------------------+---------------------------------+---------+

> create table t2 (a time(3),b datetime(3),c timestamp(3));
> show full columns from t2;
+-------+--------------+-----------+------+-----+----------------------+--------------------------------+---------------------------------+---------+
| Field | Type         | Collation | Null | Key | Default              | Extra                          | Privileges                      | Comment |
+-------+--------------+-----------+------+-----+----------------------+--------------------------------+---------------------------------+---------+
| a     | time(3)      | <null>    | YES  |     | <null>               |                                | select,insert,update,references |         |
| b     | datetime(3)  | <null>    | YES  |     | <null>               |                                | select,insert,update,references |         |
| c     | timestamp(3) | <null>    | NO   |     | CURRENT_TIMESTAMP(3) | on update CURRENT_TIMESTAMP(3) | select,insert,update,references |         |
+-------+--------------+-----------+------+-----+----------------------+--------------------------------+---------------------------------+---------+
```

Comment (xiekeyi98) [reply executor/show.go]: How about this now?
More test will be added into `tidb-test` when this PR is ready to merge.

File: executor/show.go [hunk 4]
```diff
@@ -386,6 +386,11 @@ func (e *ShowExec) fetchShowColumns() error {
 				columnDefault = defaultValStr
 			}
 		}
+		switch col.Tp {
+		case mysql.TypeTimestamp, mysql.TypeDate, mysql.TypeDuration, mysql.TypeDatetime,
+			mysql.TypeYear, mysql.TypeNewDate:
+			desc.Collation = "NULL"
```
Comment (zimulala) [inline executor/show.go]: Do we need to check other types like `int`?
```
mysql> create table t1 (a date,b datetime,c timestamp, d int);
Query OK, 0 rows affected (0.02 sec)

mysql> show full columns from t1;
+-------+-----------+-----------+------+-----+-------------------+-----------------------------+---------------------------------+---------+
| Field | Type      | Collation | Null | Key | Default           | Extra                       | Privileges                      | Comment |
+-------+-----------+-----------+------+-----+-------------------+-----------------------------+---------------------------------+---------+
| a     | date      | NULL      | YES  |     | NULL              |                             | select,insert,update,references |         |
| b     | datetime  | NULL      | YES  |     | NULL              |                             | select,insert,update,references |         |
| c     | timestamp | NULL      | NO   |     | CURRENT_TIMESTAMP | on update CURRENT_TIMESTAMP | select,insert,update,references |         |
| d     | int(11)   | NULL      | YES  |     | NULL              |                             | select,insert,update,references |         |
+-------+-----------+-----------+------+-----+-------------------+-----------------------------+---------------------------------+---------+
4 rows in set (0.01 sec)
mysql> create table t1 (a date,b datetime,c timestamp, d int);
Query OK, 0 rows affected (0.00 sec)

mysql> show full columns from t1;
+-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
| Field | Type      | Collation | Null | Key  | Default | Extra | Privileges                      | Comment |
+-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
| a     | date      | binary    | YES  |      | NULL    |       | select,insert,update,references |         |
| b     | datetime  | binary    | YES  |      | NULL    |       | select,insert,update,references |         |
| c     | timestamp | binary    | YES  |      | NULL    |       | select,insert,update,references |         |
| d     | int(11)   | binary    | YES  |      | NULL    |       | select,insert,update,references |         |
+-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
4 rows in set (0.00 sec)
```

**Comment by xiekeyi98** [comment]
I think we should file a new issue to catch it.------------------ Original ------------------From: Lynn  <notifications@github.com>Date: Wed,Mar 20,2019 8:26 PMTo: pingcap/tidb <tidb@noreply.github.com>Cc: Keyi Xie <xiekeyi98@gmail.com>, Mention <mention@noreply.github.com>Subject: Re: [pingcap/tidb] ddl:`show full columns` for date, datetime,timestamp compatibility. (#9810)@zimulala commented on this pull request.



In executor/show.go:
> @@ -386,6 +386,11 @@ func (e *ShowExec) fetchShowColumns() error {
 				columnDefault = defaultValStr
 			}
 		}
+		switch col.Tp {
+		case mysql.TypeTimestamp, mysql.TypeDate, mysql.TypeDuration, mysql.TypeDatetime,
+			mysql.TypeYear, mysql.TypeNewDate:
+			desc.Collation = "NULL"

Do we need to check other types like int?
mysql> create table t1 (a date,b datetime,c timestamp, d int);
Query OK, 0 rows affected (0.02 sec)

mysql> show full columns from t1;
+-------+-----------+-----------+------+-----+-------------------+-----------------------------+---------------------------------+---------+
| Field | Type      | Collation | Null | Key | Default           | Extra                       | Privileges                      | Comment |
+-------+-----------+-----------+------+-----+-------------------+-----------------------------+---------------------------------+---------+
| a     | date      | NULL      | YES  |     | NULL              |                             | select,insert,update,references |         |
| b     | datetime  | NULL      | YES  |     | NULL              |                             | select,insert,update,references |         |
| c     | timestamp | NULL      | NO   |     | CURRENT_TIMESTAMP | on update CURRENT_TIMESTAMP | select,insert,update,references |         |
| d     | int(11)   | NULL      | YES  |     | NULL              |                             | select,insert,update,references |         |
+-------+-----------+-----------+------+-----+-------------------+-----------------------------+---------------------------------+---------+
4 rows in set (0.01 sec)
mysql> create table t1 (a date,b datetime,c timestamp, d int);
Query OK, 0 rows affected (0.00 sec)

mysql> show full columns from t1;
+-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
| Field | Type      | Collation | Null | Key  | Default | Extra | Privileges                      | Comment |
+-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
| a     | date      | binary    | YES  |      | NULL    |       | select,insert,update,references |         |
| b     | datetime  | binary    | YES  |      | NULL    |       | select,insert,update,references |         |
| c     | timestamp | binary    | YES  |      | NULL    |       | select,insert,update,references |         |
| d     | int(11)   | binary    | YES  |      | NULL    |       | select,insert,update,references |         |
+-------+-----------+-----------+------+------+---------+-------+---------------------------------+---------+
4 rows in set (0.00 sec)


—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/pingcap/tidb","title":"pingcap/tidb","subtitle":"GitHub repository","main_image_url":"https://github.githubassets.com/images/email/message_cards/header.png","avatar_image_url":"https://github.githubassets.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/pingcap/tidb"}},"updates":{"snippets":[{"icon":"PERSON","message":"@zimulala commented on #9810"}],"action":{"name":"View Pull Request","url":"https://github.com/pingcap/tidb/pull/9810#pullrequestreview-216685680"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/pingcap/tidb/pull/9810#pullrequestreview-216685680",
"url": "https://github.com/pingcap/tidb/pull/9810#pullrequestreview-216685680",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

File: executor/show.go (hunk 4 above)
Comment (xiekeyi98) [reply executor/show.go]: I think it should be filed in another issue.

Comment (xiekeyi98) [reply executor/show.go]: https://github.com/pingcap/tidb/issues/9837
Filed the issue to trace this problem.

File: executor/show.go [hunk 5]
```diff
@@ -386,6 +386,11 @@ func (e *ShowExec) fetchShowColumns() error {
 				columnDefault = defaultValStr
 			}
 		}
+		switch col.Tp {
```
Comment (winkyao) [inline executor/show.go]: Add some comments here.

**Comment by bb7133** [comment]
LGTM

**Comment by tshqin** [comment]
need a cherry pick to 2.1

**Comment by xiekeyi98** [comment]
> need a cherry pick to 2.1

@tshqin  OK, is it emergency?
I would like to cherry pick to 2.1 when all the types have been tested.

**Pull request merged**